File: src/euler_9.py
"""Project Euler - Problem 9 Prototype."""

# Programmed by CoolCat467

from __future__ import annotations

# Project Euler - Problem 9 Prototype
# Copyright (C) 2025  CoolCat467
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received b copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

__title__ = "Project Euler - Problem 9 Prototype"
__author__ = "CoolCat467"
__license__ = "GNU General Public License Version 3"


import math
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from collections.abc import Generator


def yield_pythagorean_triplets() -> (
    Generator[tuple[int, int, int], None, None]
):
    """Yield Pythagorean Triplets.

    a < b < c && a**2 + b**2 = c**2
    """
    b = 2
    while True:
        b_2 = b * b
        for a in range(1, b):
            left = b_2 + a * a
            c = math.ceil(math.sqrt(left))
            if left == c * c:
                yield a, b, c
        b += 1


def run() -> None:
    """Run program."""
    for v in yield_pythagorean_triplets():
        if sum(v) == 1000:
            print(f"{v = }")
            print(f"{math.prod(v) = }")
            break


if __name__ == "__main__":
    print(f"{__title__}\nProgrammed by {__author__}.\n")
    run()

File: src/euler_9.rs
// Project Euler - Problem 9

// Programmed by CoolCat467

// Copyright (C) 2025  CoolCat467
//
//     This program is free software: you can redistribute it and/or modify
//     it under the terms of the GNU General Public License as published by
//     the Free Software Foundation, either version 3 of the License, or
//     (at your option) any later version.
//
//     This program is distributed in the hope that it will be useful,
//     but WITHOUT ANY WARRANTY; without even the implied warranty of
//     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//     GNU General Public License for more details.
//
//     You should have received a copy of the GNU General Public License
//     along with this program.  If not, see <https://www.gnu.org/licenses/>.

// pub fn main() {
//     let mut b: u32 = 2;
//     loop {
//         let b_2 = b * b;
//         for a in 1..b {
//             let left: u32 = b_2 + a * a;
//             let c = ((left as f64).sqrt()).ceil() as u32;
//             let right = c * c;
//             if left == right {
//                 if a + b + c == 1000 {
//                     println!("a * b * c = {}", a * b * c);
//                     return;
//                 }
//             }
//         }
//         b += 1;
//     }
// }

pub fn main() {
    for b in 2..=500 {
        for a in 1..b {
            let c = 1000 - a - b;
            // Check the Pythagorean condition
            if c > 0 && a * a + b * b == c * c {
                println!("a * b * c = {}", a * b * c);
                return;
            }
        }
    }
}
